README
Module Getopt::Long - extended processing of command line options
=================================================================

Module Getopt::Long implements an extended getopt function called
GetOptions(). This function implements the POSIX standard for command
line options, with GNU extensions, while still capable of handling
the traditional one-letter options.
In general, this means that command line options can have long names
instead of single letters, and are introduced with a double dash `--'.

Optionally, Getopt::Long can support the traditional bundling of
single-letter command line options.

Getopt::Long is part of the Perl 5 distribution. It is the successor
of newgetopt.pl that came with Perl 4. It is fully upward compatible.
In fact, the Perl 5 version of newgetopt.pl is just a wrapper around
the module.

For complete documentation, see "https://metacpan.org/release/Getopt-Long"
or use the command

    perldoc Getopt::Long

FEATURES
========

* Long option names

Major advantage of using long option names is that it is much easier
to memorize the option names. Using single-letter names one quickly
runs into the problem that there is no logical relationship between
the semantics of the selected option and its option letter.
Disadvantage is that it requires more typing. Getopt::Long provides
for option name abbreviation, so option names may be abbreviated to
uniqueness. Also, modern shells like Cornell's tcsh support option
name completion. As a rule of thumb, you can use abbreviations freely
while running commands interactively but always use the full names in
scripts. 

Examples (POSIX):

    --long --width=80 --height=24

Extensions:

    -long (convenience) +width=80 (deprecated) -height 24 (traditional)

By default, long option names are case insensitive.

* Single-letter options and bundling

When single-letter options are requested, Getopt::Long allows the
option names to be bundled, e.g. "-abc" is equivalent to "-a -b -c".
In this case, long option names must be introduced with the POSIX "--"
introducer.

Examples:

    -lgAd (bundle) -xw 80 (bundle, w takes a value) -xw80 (same)
    even -l24w80 (l = 24 and w = 80)

By default, single-letter option names are case sensitive.

* Flexibility:

  - Options can have alternative names, using an alternative name
    will behave as if the primary name was used.
  - Options can be negatable, e.g. "debug" will switch it on, while
    "nodebug" will switch it off.    
  - Options can set values, but also add values producing an array
    of values instead of a single scalar value, or set values in a hash.
  - Options can have multiple values, e.g., "--position 25 624".

* Options linkage

Using Getopt::Long gives the programmer ultimate control over the
command line options and how they must be handled:

  - By setting a global variable in the calling program.
  - By setting a specified variable.
  - By entering the option name and the value in an associative array
    (hash) or object (if it is a blessed hash).
  - By calling a user-specified subroutine with the option name and
    the value as arguments (for hash options: the name, key and value);
  - Combinations of the above.

* Customization:

The module can be customized by specifying settings in the 'use'
directive, or by calling a special method, Getopt::Long::Configure.
For example, the following two cases are functionally equal:

    use Getopt::Long qw(:config bundling no_ignore_case);

and

    use Getopt::Long;
    Getopt::Long::Configure qw(bundling no_ignore_case);

See the documentation for all possibilities.

* Object oriented interface:

Using the object oriented interface, multiple parser objects can be
instantiated, each having their own configuration settings:

    use Getopt::Long::Parser;
    $p1 = Getopt::Long::Parser->new(config => ["bundling"]);
    $p2 = Getopt::Long::Parser->(config => ["posix"]);
    if ($p1->getoptions(...options descriptions...)) ...

AVAILABILITY
============

The official version for module Getopt::Long comes with the Perl 5
distribution. 
Newer versions will be made available on the Comprehensive Perl Archive
Network (CPAN), see "https://metacpan.org/release/Getopt-Long".

Development of Gwetopt::Long is hosted on Github, 
https://github.com/sciurius/perl-Getopt-Long.
Please use the Github big tracker
https://github.com/sciurius/perl-Getopt-Long/issues
to report issues.

COPYRIGHT AND DISCLAIMER
========================

Module Getopt::Long is Copyright 1990,2013,2023 by Johan Vromans.
This program is free software; you can redistribute it and/or
modify it under the terms of the Perl Artistic License or the
GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any
later version.

-------------------------------------------------------------------
Johan Vromans                                  jvromans@squirrel.nl
Squirrel Consultancy                         Exloo, the Netherlands
https://www.squirrel.nl                   https://johan.vromans.org
------------------ "Arms are made for hugging" --------------------


CHANGES
Changes in version 2.52
-----------------------

* Fix bug https://rt.cpan.org/Ticket/History.html?id=127584.
  This is the case when an option has :s% and nothing was passed. This
  currently gives a perl warning, issue a real error message instead.

* Added some tests, see https://rt.cpan.org/Ticket/Display.html?id=87581

Changes in version 2.51
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=122068.

* Fix bug https://rt.perl.org/Ticket/Display.html?id=134352.

* Fix warnings when getting options after calling Getopt::Long->VERSION.
  Thanks to Dan Book.

* Fix issue https://rt.cpan.org/Ticket/Display.html?id=124789.

Changes in version 2.50
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=120231.

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=120300.
  Unfortunately, this withdraws a small part of fix 114999.
  GNU getopt_long() does not accept the (optional)
  argument to be passed to the option without = sign.
  We do, since not doing so breaks existing scripts.

* Provide a default value for options (gnu_compat mode).
  Thanks to Andrew Gregory.

Changes in version 2.49.1
-------------------------

* Emergency fix Makefile.PL for non-GNUmake.

Changes in version 2.49
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=114999

  Thanks Roy Ivy III for finding and fixing this.

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=113748

  Thanks Tom Wyant for finding and fixing this.

Changes in version 2.48
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=39052.

  Thanks to Roy Ivy III for digging this out and providing patches.
  
Changes in version 2.47
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=89513.

  Should have done that earlier :)

  This changes the format for accepted float numbers:

  <optional sign>
  <optional mantissa>
  <optional dec.point and fraction>
  <optional "e" and exponent>

  Mantissa and decimal fraction may not both be omitted.

  Underscores are permissible virtually anywhere in numbers for clarity.

Changes in version 2.46
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=104842

  Kudo's to EDAVIS for finding this very obscure bug.

* Some small changes/additions to the test suite and examples.

Changes in version 2.45
-----------------------

* Revert behaviour of the combination of <> and pass_through to what
  is was in 2.42.

  Thanks to Wilson Snyder to provide enhancements to the documentation
  to clarify why <> and pass_through generally is a bad idea.

  It is, however, very unfortunate to hear from users that their
  programs *depend* on the undocumented pre-2.43 behaviour.

Changes in version 2.44
-----------------------

* Be gentle to the user community and turn fix 92462 into a warning.
  This may be changed back to error in some future release.

Changes in version 2.43
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=92462

* Implement enhancement
  https://rt.cpan.org/Public/Bug/Display.html?id=101237

* Fix the Getopt::Long part of bug
  https://rt.cpan.org/Ticket/Display.html?id=100335

Changes in version 2.42
-----------------------

* Fix bug https://rt.cpan.org/Ticket/Display.html?id=88707

Changes in version 2.41
-----------------------

* Change INSTALLDIRS to site for perl >= 5.011.

  See https://rt.perl.org/rt3//Ticket/Display.html?id=116479
  for details.

Changes in version 2.40
-----------------------

* Fix bug #86095 Possible problem with aliases

* Fix bug #85362 typo fix

Changes in version 2.39
-----------------------

* Fix unneccessary warnings when @ARGV contains undefs (yes, it
  happens).

* Passing an object as first argument to the callback handler for <>
  turned out to be a problem in cases where the argument was passed to
  other modules, e.g., Archive::Tar. Revert the change since the added
  functionality of the object is not really relevant for the <>
  callback function.

* Silence the deprecation warnings from newgetopt.pl for the purpose
  of testing. These tests will be removed along with newgetopt.pl in
  the next major release of perl.
  http://perl5.git.perl.org/perl.git/commit/b814bbfa9a2087bc

* Eliminiate spurious warning.

* Retain taintedness of command line option values.

* Document that you need to check GetOptions return value :).

* Several other minor documentation fixes and enhancements.

* Fix bug  #67577 Parsing of type 'o' not correct for multiple values

Changes in version 2.38
-----------------------

* Bugfix for Ticket 35759: First arg to callback function evaluates
  to false when used in bool context.

* Fix problem with prototypes of GetOptionsFrom* functions.

* Fix restoring default die handler.

* Bugfix for Ticket 24941: Autoabbrev with + incorrect.

* Bugfix: Reject syntax 'foo|!' for option spec. This was illegal
  anyway.

Changes in version 2.37
-----------------------

* The first argument to callback function is now an object and will
  get methods for finer control in the future. The object stringifies
  to the option name, so current code should not notice a difference.

* Bugfix: With gnu_compat, --foo= will no longer trigger "Option
  requires an argument" but return the empty string.

Changes in version 2.36
-----------------------

**************** WARNING -- EXPERIMENTAL CODE AHEAD ****************

* Parsing options from an arbitrary array

  The entry point GetOptionsFromArray (exported on demand) can be used
  to parse command line options that are not passed in via @ARGV, but
  using an arbitrary array.

    use Getopt::Long qw(GetOptionsFromArray);
    $ret = GetOptionsFromArray(\@myopts, ...);

* Parsing options from an arbitrary string

  The entry point GetOptionsFromString (exported on demand) can be
  used to parse command line options that are not passed in via @ARGV,
  but using an arbitrary string.


Makefile.PL
# Makefile.PL -- Makefile generator for Getopt::Long
# Author          : Johan Vromans
# Created On      : Sat Mar 16 11:43:21 1996
# Last Modified By: Johan Vromans
# Last Modified On: Mon Aug 12 17:08:17 2019
# Update Count    : 120
# Status          : Unknown, Use with caution!

use Config;
use ExtUtils::MakeMaker 5.0;

use lib "./lib";
use Getopt::Long;
my $version = "2.51";

my $mm_version = $ExtUtils::MakeMaker::VERSION;
if ($mm_version =~ /_/) {
    # developer release
    $mm_version = eval $mm_version;
    die $@ if $@;
}

sub MY::postamble {
    my $ret = "";
    my $mandir = $Config{installman3dir};
    my $obsolete = "$mandir/Getopt::GetoptLong.3";
    if ( -e $obsolete ) {
	print STDERR (<<EOD);

I found a left-over manual page from a previous version of Getopt::Long.
This manual page has a wrong name, and should be removed.
I can add instructions to the Makefile to remove it when you
install this version.
EOD
	my $ans = prompt ("Shall I add the remove instructions?");
	if ( $ans =~ /^y/i ) {
	    $ret .= <<EOD;
install ::
	\$(RM_F) "$obsolete"

EOD
	}
    }
    my $libdir = $Config{installprivlib};
    my $obsolete = "$libdir/auto/Getopt/Long";
    if ( -d $obsolete ) {
	print STDERR (<<EOD);

I found left-over files from a previous version of Getopt::Long.
These should be removed.
I can add instructions to the Makefile to remove them when you
install this version.
EOD
	my $ans = prompt ("Shall I add the remove instructions?");
	if ( $ans =~ /^y/i ) {
	    $ret .= <<EOD;
install ::
	\$(RM_RF) "$libdir/auto/Getopt/Long"
EOD
	}
    }
    $ret;
}

if ( $version =~ /_/ ) {
    print STDERR <<EOD;

WARNING: This is Getopt::Long version $version.
The underscore in the version number indicates that this is
an intermediate, preview or beta release.
Please keep checking CPAN for the official release.
EOD
}

unless ( eval { require Pod::Usage } ) {
    print STDERR <<EOD;

WARNING: Getopt::Long needs the module Pod::Usage for its built-in
help facility. Pod::Usage seems not to be installed on this system.
That is not a problem as long as the new help facility is not used.
However, please consider installing Pod::Usage from CPAN as soon as
convenient.
EOD
}

print STDERR ("\n");

WriteMakefile(
	NAME      => "Getopt::Long",
	VERSION   => $version,

	( $] < 5.005 ) ? () : (
	AUTHOR => 'Johan Vromans (jvromans@squirrel.nl)',
        ABSTRACT => 'Module to handle parsing command line options',
        ),

	INSTALLDIRS => ($] < 5.011 ? 'perl' : 'site'),
	PM 	  => { "lib/newgetopt.pl"  => '$(INST_LIB)/newgetopt.pl',
		       "lib/Getopt/Long.pm" => '$(INST_LIB)/Getopt/Long.pm',
		     },
	dist      => { COMPRESS => 'gzip', SUFFIX => 'gz',
		     },
	PREREQ_PM => { "Pod::Usage" => 1.14 },

        ($mm_version <= 6.45) ? () : (
            META_MERGE => {
                'meta-spec' => { version => 2 },
                resources => {
                    repository => {
                        type => "git",
                        url => "https://github.com/sciurius/perl-Getopt-Long.git",
                        web => "https://github.com/sciurius/perl-Getopt-Long",
                    },
                },
            },
        ),


 );

